packages/next-on-pages/templates/cache/interface.ts

```diff
@@ -0,0 +1,244 @@
+const revalidatedTags = new Set<string>();
+
+/** Generic interface for the Suspense Cache. */
+export class CacheInterface {
+	/** The tags manifest for fetch calls. */
+	public tagsManifest: TagsManifest | undefined;
+	/** The key used for the tags manifest in the cache. */
+	public tagsManifestKey = 'tags-manifest';
+
+	/**
+	 * @param ctx The incremental cache context.
+	 */
+	constructor(protected ctx: Record<string, unknown> = {}) {}
+
+	/**
+	 * Retrieves an entry from the storage mechanism.
+	 *
+	 * @param key Key for the item.
+	 * @returns The value, or null if no entry exists.
+	 */
+	public async retrieve(key: string): Promise<string | null> {
+		throw new Error(`Method not implemented - ${key}`);
+	}
+
+	/**
+	 * Updates an entry in the storage mechanism.
+	 *
+	 * @param key Key for the item.
+	 * @param value The value to update.
+	 */
+	public async update(key: string, value: string): Promise<void> {
+		throw new Error(`Method not implemented - ${key}, ${value}`);
+	}
+
+	/**
+	 * Puts a new entry in the suspense cache.
+	 *
+	 * @param key Key for the item in the suspense cache.
+	 * @param value The cached value to add to the suspense cache.
+	 */
+	public async set(key: string, value: IncrementalCacheValue): Promise<void> {
+		const newEntry: CacheHandlerValue = {
+			lastModified: Date.now(),
+			value,
+		};
+
+		// Update the cache entry.
+		await this.update(key, JSON.stringify(newEntry));
+
+		switch (newEntry.value?.kind) {
+			case 'FETCH': {
+				// Update the tags with the cache key.
+				const tags = newEntry.value.data.tags ?? [];
+				await this.setTags(tags, { cacheKey: key });
+
+				getDerivedTags(tags).forEach(tag => revalidatedTags.delete(tag));
+			}
+		}
+	}
+
+	/**
+	 * Retrieves an entry from the suspense cache.
+	 *
+	 * @param key Key for the item in the suspense cache.
+	 * @returns The cached value, or null if no entry exists.
+	 */
+	public async get(key: string): Promise<CacheHandlerValue | null> {
+		// Get entry from the cache.
+		const entry = await this.retrieve(key);
+		if (!entry) return null;
+
+		let data: CacheHandlerValue;
+		try {
+			data = JSON.parse(entry) as CacheHandlerValue;
+		} catch (e) {
+			// Failed to parse the cache entry, so it's invalid.
+			return null;
+		}
+
+		switch (data.value?.kind) {
+			case 'FETCH': {
+				// Load the tags manifest.
+				await this.loadTagsManifest();
+
+				// Check if the cache entry is stale or fresh based on the tags.
+				const tags = getDerivedTags(data.value.data.tags ?? []);
+				const isStale = tags.some(tag => {
+					// If a revalidation has been triggered, the current entry is stale.
+					if (revalidatedTags.has(tag)) return true;
+
+					const tagEntry = this.tagsManifest?.items?.[tag];
+					return (
+						tagEntry?.revalidatedAt &&
+						tagEntry?.revalidatedAt >= (data.lastModified ?? Date.now())
+					);
+				});
+
+				// Don't return stale data from the cache.
+				return isStale ? null : data;
```

Comment:
I think we discussed this and that even state data should be returned?

Reply:
This part is different to what we discussed. Here, it is considered stale and returning null if a revalidation is triggered and the data is still old. What we discussed was returning data that is old but has not had a revalidation triggered for it yet.

I would also note that this is consistent with how Next.js' file system cache works - https://github.com/vercel/next.js/blob/baef523fb08e1a7d588026ba784247ebcfd07740/packages/next/src/server/lib/incremental-cache/file-system-cache.ts#L277

Reply:
ahhh ok I see! ok, thanks! 🙏